File: LinearHashing.ts
import { PageList, Page } from "./PageList";

export default class LinearHashing {
  private pages: Map<number, PageList>;
  private N: number = 0;
  private l: number = 0;
  public elementsCount: number;
  public pagesCount: number;
  constructor(
    private p: number = 2,
    private maxAlpha: number = 0.8,
    private m: number = 2
  ) {
    this.pages = new Map<number, PageList>();
    for (let i = 0; i < this.m; i++) {
      this.pages.set(i, new PageList(this.p));
    }

    this.pagesCount = this.m;
    this.elementsCount = 0;
  }

  private hash(value: number, level: number) {
    return value % (2 ** level * this.m);
  }

  private insertInPage(index: number, value: number) {
    this.pagesCount += this.pages.get(index)!.insert(value);
  }

  public insert(value: number) {
    let index = this.hash(value, this.l);
    if (index < this.N) {
      index = this.hash(value, this.l + 1);
    }

    this.insertInPage(index, value);
    this.elementsCount++;

    while (this.alpha > this.maxAlpha) {
      this.reorder();
    }
  }

  private reorder() {
    this.pages.set(this.N + 2 ** this.l * this.m, new PageList(this.p));

    const pageList = this.pages.get(this.N)!;
    let currentPage: Page | null = pageList.head;
    let stayingKeys: number[] = [];
    let keysCount: number = 0;
    while (currentPage) {
      for (let i = 0; i < currentPage.values.length; i++) {
        let newIndex = this.hash(currentPage.values[i], this.l + 1);
        if (newIndex == this.N) {
          stayingKeys.push(currentPage.values[i]);
        } else {
          this.pages.get(newIndex)!.insert(currentPage.values[i]);
        }
        keysCount++;
      }
      currentPage = currentPage.next;
    }

    // count the number of pages added in total after reordering (this will always be either 1 or 0)
    this.pagesCount +=
      Math.ceil(Math.max(stayingKeys.length, 1) / this.p) +
      Math.ceil(Math.max(keysCount - stayingKeys.length, 1) / this.p) -
      Math.ceil(keysCount / this.p);

    const newPageList = new PageList(this.p);
    for (let i = 0; i < stayingKeys.length; i++) {
      newPageList.insert(stayingKeys[i]);
    }
    this.pages.set(this.N, newPageList);

    this.N++;
    if (this.N >= 2 ** (this.l + 1)) {
      this.l++;
      this.N = 0;
    }
  }

  public get alpha():number{
    return this.elementsCount / (this.pagesCount * this.p)
  }

  public get pStar(){
    return this.pagesCount / this.pages.size
  }

  public print() {
    for (let i = 0; i < this.pages.size; i++) {
      console.log(`Page ${i}: `);
      this.pages.get(i)!.print();
    }
  }
}

// const linearHashing = new LinearHashing(2, 0.8);
// linearHashing.insert(16)
// linearHashing.insert(15)
// linearHashing.insert(20)
// linearHashing.insert(36)
// linearHashing.insert(48)
// linearHashing.insert(40)
// linearHashing.insert(64)
// linearHashing.insert(88)
// linearHashing.insert(96)
// linearHashing.insert(3)
// linearHashing.print();

// const linearHashing = new LinearHashing(2, 0.8);
// linearHashing.insert(8);
// linearHashing.insert(11);
// linearHashing.insert(10);
// linearHashing.insert(15);
// linearHashing.insert(17);
// linearHashing.insert(25);
// linearHashing.insert(44);
// linearHashing.insert(12)

// linearHashing.print();

// const linearHashing = new LinearHashing(2, 0.8);
// linearHashing.insert(8);
// linearHashing.insert(17);
// linearHashing.insert(25);
// linearHashing.insert(10);
// linearHashing.insert(11);
// linearHashing.insert(15);
// linearHashing.insert(44);
// linearHashing.insert(12);
// linearHashing.insert(22);
// linearHashing.insert(37);
// linearHashing.insert(30);
// linearHashing.insert(21);

// linearHashing.print();

File: src/linear_hashing.py
from math import ceil
from .page_list import PageList


class LinearHashing:
    def __init__(self, p=2, max_alpha=0.8, m=2):
        self.pages = {}
        self.N = 0
        self.l = 0
        self.elements_count = 0
        self.pages_count = 0
        self.p = p
        self.max_alpha = max_alpha
        self.m = m

        for i in range(self.m):
            self.pages[i] = PageList(self.p)

        self.pages_count = self.m
        self.elements_count = 0

    def hash(self, value, level):
        return value % (2**level * self.m)

    def insert_in_page(self, index, value):
        self.pages_count += self.pages[index].insert(value)

    def insert(self, value):
        index = self.hash(value, self.l)
        if index < self.N:
            index = self.hash(value, self.l + 1)

        self.insert_in_page(index, value)
        self.elements_count += 1

        while self.alpha > self.max_alpha:
            self.reorder()

    def reorder(self):
        self.pages[self.N + 2**self.l * self.m] = PageList(self.p)

        page_list = self.pages[self.N]
        current_page = page_list.head
        staying_keys = []
        keys_count = 0

        while current_page:
            for val in current_page.values:
                new_index = self.hash(val, self.l + 1)
                if new_index == self.N:
                    staying_keys.append(val)
                else:
                    self.pages[new_index].insert(val)
                keys_count += 1

            current_page = current_page.next

        self.pages_count += (
            ceil(max(len(staying_keys), 1) / self.p)
            + ceil(max(keys_count - len(staying_keys), 1) / self.p)
            - ceil(keys_count / self.p)
        )

        new_page_list = PageList(self.p)

        for key in staying_keys:
            new_page_list.insert(key)
        self.pages[self.N] = new_page_list

        self.N += 1
        if self.N >= 2 ** (self.l + 1):
            self.l += 1
            self.N = 0

    @property
    def alpha(self):
        return self.elements_count / (self.pages_count * self.p)

    @property
    def p_star(self):
        return self.pages_count / len(self.pages)

    def print(self):
        for i in range(len(self.pages)):
            print(f"Page {i}:")
            self.pages[i].print()
